pkg/cloud/openstack/clients/secgroupservice.go

```diff
@@ -0,0 +1,358 @@
+	return openstackconfigv1.SecurityGroup{
+		Name: secGroupName,
+		Rules: append(
+			[]openstackconfigv1.SecurityGroupRule{
+				{
+					Direction:      "ingress",
+					EtherType:      "IPv4",
+					PortRangeMin:   443,
+					PortRangeMax:   443,
+					Protocol:       "tcp",
+					RemoteIPPrefix: "0.0.0.0/0",
+				},
+				{
+					Direction:      "ingress",
+					EtherType:      "IPv4",
+					PortRangeMin:   22,
+					PortRangeMax:   22,
+					Protocol:       "tcp",
+					RemoteIPPrefix: "0.0.0.0/0",
+				},
+			},
+			defaultRules...,
+		),
+	}
+}
+
+func (s *SecGroupService) generateGlobalGroup(clusterName string) openstackconfigv1.SecurityGroup {
+	secGroupName := fmt.Sprintf("%s-cluster-%s-secgroup-%s", secGroupPrefix, clusterName, globalPostfix)
+
+	// As above, hardcoded rules.
+	return openstackconfigv1.SecurityGroup{
+		Name: secGroupName,
+		Rules: append(
+			[]openstackconfigv1.SecurityGroupRule{
+				{
+					Direction:     "ingress",
+					EtherType:     "IPv4",
+					PortRangeMin:  1,
+					PortRangeMax:  65535,
+					Protocol:      "tcp",
+					RemoteGroupID: "self",
+				},
+				{
+					Direction:     "ingress",
+					EtherType:     "IPv4",
+					PortRangeMin:  1,
+					PortRangeMax:  65535,
+					Protocol:      "udp",
+					RemoteGroupID: "self",
+				},
+				{
+					Direction:     "ingress",
+					EtherType:     "IPv4",
+					PortRangeMin:  0,
+					PortRangeMax:  0,
+					Protocol:      "icmp",
+					RemoteGroupID: "self",
+				},
+			},
+			defaultRules...,
+		),
+	}
+}
+
+// matchGroups will check if security groups match.
+func (s *SecGroupService) matchGroups(desired, observed *openstackconfigv1.SecurityGroup) bool {
+	// If they have differing amount of rules they obviously don't match.
+	if len(desired.Rules) != len(observed.Rules) {
+		return false
+	}
+
+	// Rules aren't in any order, so we're doing this the hard way.
+	for _, desiredRule := range desired.Rules {
+		r := desiredRule
+		if r.RemoteGroupID == "self" {
+			r.RemoteGroupID = observed.ID
+		}
+		ruleMatched := false
+		for _, observedRule := range observed.Rules {
+			if s.matchRule(r, observedRule) {
+				ruleMatched = true
+				break
+			}
+		}
+
+		if !ruleMatched {
+			return false
+		}
+	}
+	return true
+}
+
+// reconcileGroup reconciles an already existing observed group by essentially emptying out all the rules and
+// recreating them.
+func (s *SecGroupService) reconcileGroup(desired, observed *openstackconfigv1.SecurityGroup) (*openstackconfigv1.SecurityGroup, error) {
+	klog.V(6).Infof("Deleting all rules for group %s", observed.Name)
+	for _, rule := range observed.Rules {
+		klog.V(6).Infof("Deleting rule %s from group %s", rule.ID, observed.Name)
+		err := rules.Delete(s.client, rule.ID).ExtractErr()
+		if err != nil {
+			return &openstackconfigv1.SecurityGroup{}, err
+		}
+	}
+	recreatedRules := make([]openstackconfigv1.SecurityGroupRule, len(desired.Rules), len(desired.Rules))
```

Comment:
`make([]..., 3, 3)` is redundant. `make([]..., 3)`. Both create a slice with `cap = len = 3`. Which means, if you do `append`, you will increase the slice, and not fill from zero.

As the code looks like, I think what you wanted to make is `make([...], 0, len(x))`.
There are some more places with the same pattern in this file. See https://play.golang.org/p/dnmg_tce2U5